### DISCUSSION TITLE
Can I help you with the Xbox Elite V2 Controller?

### DISCUSSION THREAD

**Discussion by iMonZ** [ndeadly/MissionControl#317] [Ideas]
Hey i have an Xbox Elite V2 Controller and would like to help you debugging it.
I have basic C++ and C knowledge too

**Comment by ndeadly** [comment]
While C/C++ knowledge would help, in order to contribute anything meaningful here you would also need some understanding of ARM64 assembly, using an RE tool like IDA or Ghidra and debugging running code with gdb. Basically we need to debug the running bluetooth service (for which we don't have the source code) while attempting to pair the controller and try to figure out where and why the connection is being terminated prematurely, with the hope that it's possible to patch the binary in some way to fix/work around the issue.

I'd certainly welcome your contributions if you think this is something you can handle, and I'd be happy to give you some pointers, but if you lack the skills mentioned above you're going to need a while to develop them first before you can think about attacking the actual  problem.

**Reply by iMonZ** [reply]
> While C/C++ knowledge would help, in order to contribute anything meaningful here you would also need some understanding of ARM64 assembly, using an RE tool like IDA or Ghidra and debugging running code with gdb. Basically we need to debug the running bluetooth service (for which we don't have the source code) while attempting to pair the controller and try to figure out where and why the connection is being terminated prematurely, with the hope that it's possible to patch the binary in some way to fix/work around the issue.
> 
> I'd certainly welcome your contributions if you think this is something you can handle, and I'd be happy to give you some pointers, but if you lack the skills mentioned above you're going to need a while to develop them first before you can think about attacking the actual problem.

C✅
C++ ✅
Ghidra (I used this tool just once)
And I don’t have assembly knowledge.

Do you have some books you can recommend?

**Reply by ndeadly** [reply]
I personally never read any books on the matter, just dove in with some experience in C/C++ and MIPS assembly from university.  I read a few IDA tutorials or maybe watched a youtube video or two to help me get started with finding my way around the RE tool, but there was nothing to be found that really jumped out at me as being a great resource.

I think most people involved with RE will probably tell you that you can read all the books you want, but the real learning comes from just getting hands on with the software and exploring compiled binaries for yourself to get a feel for how source code is translated to assembly by the compiler, how a program is laid out in memory. how giving the software hints about data types and things can improve the decompiled pseudocode etc.

You can definitely get pretty far without knowing how to program assembly, but a basic understanding of the types of instructions used by a typical CPU and how they operate on values stored in registers and memory would be beneficial to aid in understanding, and pretty much necessary if you're going to be looking at patching instructions to change the behaviour of the binary.

I don't want to discourage you if you're willing to learn, but this isn't really something you'll pick up in a couple of evenings. You'll probably need weeks or even months of messing about with it before you start to feel you have even a remote idea that you know what you're doing. I still feel I don't know what I'm doing sometimes...
